### PR TITLE
[DEV-4645] Multiple CFDA downloads

### PR DESCRIPTION
**Pull request by emily-brents**
**Description:**
Includes all CFDAs associated with an award in the download, not just the CFDA associated with the latest transaction

**Technical details:**
Changes `cfda_title` and `cfda_number` to `cfda_numbers_and_titles` in the award download and `prime_award_cfda_number` and `prime_award_cfda_title` to `prime_award_cfda_numbers_and_titles` in the subgrant download

**Requirements for PR merge:**

1. [X] Unit & integration tests updated
2. [X] API documentation updated
3. [ ] Necessary PR reviewers:
    - [ ] Backend
4. [X] Matview impact assessment completed
5. [X] Frontend impact assessment completed
6. [X] Data validation completed
7. [X] Appropriate Operations ticket(s) created
8. [X] Jira Ticket [DEV-4645](https://federal-spending-transparency.atlassian.net/browse/DEV-4645):
    - [ ] Link to this Pull-Request
    - [ ] Performance evaluation of affected (API | Script | Download)
    - [ ] Before / After data comparison

**Area for explaining above N/A when needed:**
```
```

### CODE REVIEW

File: usaspending_api/download/v2/download_column_historical_lookups.py [hunk 1]
```diff
@@ -716,8 +716,7 @@
                     "primary_place_of_performance_foreign_location",
                     "award__latest_transaction__assistance_data__place_of_performance_forei",
                 ),
-                ("cfda_number", "award__latest_transaction__assistance_data__cfda_number"),
-                ("cfda_title", "award__latest_transaction__assistance_data__cfda_title"),
+                ("cfda_numbers_and_titles", None),  # Annotation is used to create this column
```
Comment (kwhickey) [inline usaspending_api/download/v2/download_column_historical_lookups.py:719]: Did requirements get clarified on the story to _combine_ what was 2 columns into 1? Originally I think Ross had it written out as keeping 2 cols, and just ordering the numbers and titles in each column the same so they "lined up".

File: usaspending_api/download/helpers/download_annotation_functions.py [hunk 2]
```diff
@@ -563,5 +579,21 @@ def subaward_annotations():
...
+            TransactionFABS.objects.filter(transaction__award_id=OuterRef("award_id"))
+            .values("cfda_number")
+            .annotate(
+                value=ExpressionWrapper(
+                    ConcatAll(F("cfda_number"), Value(": "), F("cfda_title")),
+                    output_field=TextField(),
+                ),
+                sum=Sum(Coalesce("federal_action_obligation", "original_loan_subsidy_cost")),
+            )
+            .order_by("sum")
```
Comment (kwhickey) [inline usaspending_api/download/helpers/download_annotation_functions.py:592]: Was this sum put there _just_ for determining the sort-order of the CFDAs listed in the row's CFDA column(s)? 

If so, were you able to test out the impact of this extra mini-agg on downloads of really large files covering a lot of awards (esp. those that may have many CFDAs in their transactions?). I think this extra bit of usability (sorting by "predominant" CFDA) should only be added if it has nearly zero impact on performance.

File: usaspending_api/download/helpers/download_annotation_functions.py [hunk 3]
```diff
@@ -266,6 +266,22 @@ def universal_award_matview_annotations():
             output_field=TextField(),
         ),
         "award_latest_action_date_fiscal_year": FiscalYear(F("award__latest_transaction__action_date")),
+        "cfda_numbers_and_titles": Subquery(
+            TransactionFABS.objects.filter(transaction__award_id=OuterRef("award_id"))
+            .values("cfda_number")
+            .annotate(
```
Comment (kwhickey) [inline usaspending_api/download/helpers/download_annotation_functions.py:272]: Same question as below around sum and performance impact.

**Review by kwhickey — COMMENTED** [review]
Nice and clean but a few questions on column combining and performance impact of sorting the CFDAs by total amount.

File: usaspending_api/download/v2/download_column_historical_lookups.py (hunk 1 above)
Comment (emily-brents) [reply usaspending_api/download/v2/download_column_historical_lookups.py:719]: There wasn't really any clarification on which method was best, I personally think it works better UX-wise and with the fact that we have DEFC, program activity, and object class the same way having the titles and numbers together, but it's trivial to separate them so I'll change it if that makes more sense to people.

File: usaspending_api/download/helpers/download_annotation_functions.py (hunk 2 above)
Comment (emily-brents) [reply usaspending_api/download/helpers/download_annotation_functions.py:592]: Yeah, the AC wants them sorted by which CFDA has the most money associated with it per award. 

I'm running some test with larger data sets right now, but at the moment it looks like there is some performance lag - 41s without the sum v. 50s with the sum.

Comment (sethstoudenmier) [reply usaspending_api/download/helpers/download_annotation_functions.py]: Are the tests against local or a NonProd environment?

File: usaspending_api/disaster/management/sql/disaster_covid19_file_d2_awards.sql [hunk 4]
```diff
@@ -64,9 +64,7 @@ SELECT
     "transaction_fabs"."place_of_performance_zip4a" AS "primary_place_of_performance_zip_4",
     "transaction_fabs"."place_of_performance_congr" AS "primary_place_of_performance_congressional_district",
     "transaction_fabs"."place_of_performance_forei" AS "primary_place_of_performance_foreign_location",
-    "transaction_fabs"."cfda_number" AS "cfda_number",
-    "transaction_fabs"."cfda_title" AS "cfda_title",
-    "transaction_fabs"."assistance_type" AS "assistance_type_code",
+    (SELECT STRING_AGG(DISTINCT CONCAT(U0."cfda_number", ': ', U0."cfda_title"), '; ' ) AS "total" FROM "transaction_fabs" U0 INNER JOIN "transaction_normalized" U1 ON (U0."transaction_id" = U1."id") WHERE U1."award_id" = ("vw_award_search"."award_id") GROUP BY U1."award_id") AS "cfda_numbers_and_titles",    "transaction_fabs"."assistance_type" AS "assistance_type_code",
```
Comment (sethstoudenmier) [inline usaspending_api/disaster/management/sql/disaster_covid19_file_d2_awards.sql:67]: `"transaction_fabs"."assistance_type" AS "assistance_type_code",` should be on a different line.

Also, `vw_award_search` is never referenced in a `from` so this is blowing up the query. Did you means `"awards"."id`?

File: usaspending_api/disaster/management/sql/disaster_covid19_file_f_grants.sql [hunk 5]
```diff
@@ -52,9 +52,7 @@ SELECT
     "broker_subaward"."place_of_perform_country_co" AS "prime_award_primary_place_of_performance_country_code",
     "broker_subaward"."place_of_perform_country_na" AS "prime_award_primary_place_of_performance_country_name",
     "broker_subaward"."award_description" AS "prime_award_description",
-    "broker_subaward"."cfda_numbers" AS "prime_award_cfda_number",
-    "broker_subaward"."cfda_titles" AS "prime_award_cfda_title",
-    "broker_subaward"."subaward_type" AS "subaward_type",
+    (SELECT STRING_AGG(DISTINCT CONCAT(U0."cfda_number", ': ', U0."cfda_title"), '; ' ) AS "total" FROM "transaction_fabs" U0 INNER JOIN "transaction_normalized" U1 ON (U0."transaction_id" = U1."id") WHERE U1."award_id" = ("subaward_view"."award_id") GROUP BY U1."award_id") AS "prime_award_cfda_numbers_and_titles",    "broker_subaward"."subaward_type" AS "subaward_type",
```
Comment (sethstoudenmier) [inline usaspending_api/disaster/management/sql/disaster_covid19_file_f_grants.sql:55]: Similar questions as above

File: usaspending_api/download/helpers/download_annotation_functions.py (hunk 2 above)
Comment (emily-brents) [reply usaspending_api/download/helpers/download_annotation_functions.py]: tested against QAT db but running the API locally

File: usaspending_api/disaster/management/sql/disaster_covid19_file_d2_awards.sql (hunk 4 above)
Comment (emily-brents) [reply usaspending_api/disaster/management/sql/disaster_covid19_file_d2_awards.sql]: should be fixed

**Review by sethstoudenmier — DISMISSED** [review]
LGTM. Just need to confirm we are okay to not have sorting by amount before merging.

**Review by sethstoudenmier — APPROVED** [review]
LGTM

**Pull request merged**